config: add support for control plane identifier

ramaraochavali opened the pull request:
Description: Adds support for specifying control plane identifier in Discovery Response and displays it in `/configdump`( we initially discussed to display in `/configdump` but later thought it would be as text stat since text stat is taking time I am currently displaying it in `/configdump' and once the text stat is available, will publish it as text stat also).
Risk Level: Low
Testing: Added automated tests
Docs Changes: Added
Release Notes: Updated
Fixes #5770

ramaraochavali commented:
@fredlas No problem. There are couple of issues with text stat -
1. Its support in Envoy - which I think can be fixed.
2. Its support in stats backends - I see that prometheus does not support so as the Metrics.proto we use for Metrics service. I have not checked if statsd supports it or not. So even if we support in Envoy, not sure how many will be able to use it. That is why I am adding it to config dump as well.

ramaraochavali commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -206,5 +217,13 @@ void GrpcMuxImpl::handleEstablishmentFailure() {
 
+ProtobufTypes::MessagePtr GrpcMuxImpl::dumpControlPlaneConfig() const {
+  auto config_dump = std::make_unique<envoy::admin::v2alpha::ControlPlaneConfigDump>();
+  config_dump->mutable_control_plane()->set_identifier(control_plane_.control_plane_identifier_);
```
yes.

ramaraochavali commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -12,10 +14,13 @@ GrpcMuxImpl::GrpcMuxImpl(const LocalInfo::LocalInfo& local_info, Grpc::AsyncClie
           std::move(async_client), service_method, random, dispatcher, scope, rate_limit_settings),
-      local_info_(local_info) {
+      local_info_(local_info), config_tracker_entry_(config_tracker.add(
+                                   "control_plane", [this] { return dumpControlPlaneConfig(); })),
```
I do not think that is possible with out ConfigTracker machinery. You need to register this to tracker then only it calls when `/config_dump' is accessed. May be @mattklein123  knows if there is any other way.

mattklein123 replied:
The config tracker could be registered elsewhere, and there could be separate machinery to keep track of this state. Will this even work though for multiple xDS connections though? Won't this get registered multiple times if the user is not using ADS? I think this probably needs some rethinking?

ramaraochavali replied:
I thought about multiple xDS implementations. My thought was that it works similar to how `control_plane.connected_state` stat is working. Regarding duplicate, I think based on what I saw in config_tracker only one will be registered, others will not be added - Is that not correct?

mattklein123 replied:
I'm not sure, you would need to look at the code. As I recall it will not register multiple times. Also, I don't think last one wins here makes sense. We would want to print the information for all connections?

ramaraochavali replied:
I am fine with that. I can change it to print with all connections and register them multiple times prefixing/suffixing the xDS it is registering. Is that fine?

mattklein123 replied:
I would have to think it through, don't have a great answer for you. :) As long as print for all connections I think it should be fine. I would try to have an integration test for this. I suspect it wouldn't be that hard given some of the other xDS and config dump tests we have.

ramaraochavali replied:
@mattklein123  I thought through a littile bit and I think most easy and generic way is to build `ControlPlaneConfigDump` per each service like below. So I ADS is used, only first one will come. If user uses individual services, it gets repeated for each service - Do you think it is fine? If yes, I will make changes accordingly 

```
  {
   "@type": "type.googleapis.com/envoy.admin.v2alpha.ControlPlaneConfigDump",
     "service": "envoy.service.discovery.v2.AggregatedDiscoveryService",
     "control_plane": {
	    "identifier" : "control_plane_1"
	 },
     "last_updated": "1970-01-01T00:00:00Z"
  },
  {
   "@type": "type.googleapis.com/envoy.admin.v2alpha.ControlPlaneConfigDump",
     "service": "envoy.api.v2.EndpointDiscoveryService",
     "control_plane": {
	    "identifier" : "control_plane_1"
	 },
     "last_updated": "1970-01-01T00:00:00Z"
  }
```

mattklein123 replied:
But I think it's possible there are multiple connections for each service? Let's say you have 2 services, each using an independent RDS connection? I think that is possible currently? (I can't actually remember if we multiplex all RDS over a single xDS connection, we might.)

ramaraochavali replied:
Sorry . Did follow fully - Are you saying that different Listeners can provide different RDS config sources?

mattklein123 replied:
I think so but I can't remember. I would have a look. cc @AndresGuedez

htuch replied:
Yes, independent listeners can have independent RDS config source and connections.

mattklein123 commented on api/envoy/admin/v2alpha/config_dump.proto:
```diff
@@ -29,8 +30,11 @@ message ConfigDump {
+  // <envoy_api_msg_admin.v2alpha.ControlPlaneConfigDump>`
   // * *listeners*: :ref:`ListenersConfigDump <envoy_api_msg_admin.v2alpha.ListenersConfigDump>`
   // * *routes*:  :ref:`RoutesConfigDump <envoy_api_msg_admin.v2alpha.RoutesConfigDump>`
+
```
Check the rendered docs, I'm guessing you don't want this new line.

htuch commented on api/envoy/admin/v2alpha/config_dump.proto:
```diff
@@ -178,3 +182,12 @@ message RoutesConfigDump {
+// This message describes the control plane configuration that Envoy is currently connected to.
+message ControlPlaneConfigDump {
+  // The control plane that Envoy is currently connected to.
+  envoy.api.v2.core.ControlPlane control_plane = 1 [(gogoproto.nullable) = false];
```
If we set `gogoproto` constraints here, shouldn't we set the PGV equivalent?

htuch commented on source/common/config/grpc_mux_impl.h:
```diff
@@ -85,10 +87,19 @@ class GrpcMuxImpl
   };
 
+  // ControlPlane with which Envoy is connected to.
+  struct ControlPlane {
```
In this case, you could just stick with the protobuf message type I think.

ramaraochavali commented:
@htuch if we want to view all config sources as unordered set, I guess the way is to maintain that static set in grpc_mux_impl - is that fine? Or is there any better way? Or we could pass config source to grpc_mux_impl. But we cannot stitch all of them together as a set - we have display them independently. Let me know if there is a better way

htuch commented:
@ramaraochavali I think `ConfigTracker` is the right approach, you just need to allow for multiple entries for each xDS there.

ramaraochavali commented:
@htuch Sorry for the delay here. I change the proto a bit to support multiple config sources per service and changed the grpc_mux_impl accordingly. If this is what we want and is in the right direction, I will add few more tests to validate possible combinations.

htuch commented on api/envoy/admin/v2alpha/config_dump.proto:
```diff
@@ -178,3 +182,38 @@ message RoutesConfigDump {
+  message ConfigSourceControlPlaneInfo {
+
+    // gRPC service represented by ApiConfigSource.
+    envoy.api.v2.core.GrpcService grpc_service = 1;
```
I'm not sure if this provides unique identification. If you have lets say two clusters X and Y which uses EDS. IIUC you then have one `ServiceControlPlaneInfo` for EDS and then two `ConfigSourceControlPlaneInfo` inside that. They may point at the same EDS gRPC service but might have separate connections and last updates.

The way things work together, in the above example, they are muxed, but this is really an implementation detail. I'd like to support the idea that you could split different EDS subsets to the same control plane over multiple connections if non-aggregated.

What's missing is the notion of the CDS resource that depends on the EDS object here. This is pretty hard to capture generically, maybe we could have a field that is `parent` that names the resource responsible for creating the xDS request. For CDS/LDS, this might be empty, for SDS/LDS/RDS, this could reflect the `Listener` or `Cluster` responsible.

WDYT? @fredlas also for thinking on this.

fredlas replied:
I don't quite understand. If each cluster gets its own `ConfigSourceControlPlaneInfo`, even if duplicate, doesn't that already support the goal of being agnostic to whether they're being aggregated over the same gRPC channel or even stream?

As for the dependencies... actually, I think I don't entirely understand most of what's going on here. I thought the idea was to just report "whom" Envoy's xDS connections are going to. I would expect that to be either just a bunch of ControlPlane strings, or a bunch of pairs of (ControlPlane string, type URL string).

ramaraochavali replied:
@htuch
>>  IIUC you then have one ServiceControlPlaneInfo for EDS and then two ConfigSourceControlPlaneInfo inside that. They may point at the same EDS gRPC service but might have separate connections and last updates.

In that case I was thinking that there will be only be one `ConfigSourceControlPlaneInfo` inside `ServiceControlPlaneInfo` - That is how the implementation was. It checks if there is a `ConfigSourceControlPlaneInfo` for a gRPC service - it just updates the existing instead of creating a new one. 

Are you thinking we should represent them as two entries?

htuch replied:
I guess it depends on what you want to do with the information. Do you want to just know who you are connected to coarsely, or do you want to know details on how configuration is being delivered for a specific cluster or listener?

ramaraochavali replied:
@htuch  My main intent was to  to just know who you are connected to coarsely - to which control Envoy is connected.

htuch replied:
If it's only "I'm connected to this set of servers for these gRPC services", then what you have is fine. Thanks for clarifying.

htuch commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -206,5 +220,45 @@ void GrpcMuxImpl::handleEstablishmentFailure() {
   }
 }
 
+void GrpcMuxImpl::populateControlPlaneInfo(const envoy::api::v2::DiscoveryResponse& message) {
```
FWIW, there is a bunch of refactoring going on here in https://github.com/envoyproxy/envoy/pull/6430 and followup PRs. Might be some fun merge issues :)

ramaraochavali commented:
@fredlas 
>> I realized: it looks like only gRPC xDS is supported here. If nobody who uses REST/filesystem/whatever wants this, then that makes sense.

Yes. That was intentional atleast from my side to just support it for gRPC - and if is useful and want to extend it for other subscriptions, I can do a follow-up

htuch commented:
@ramaraochavali let me know if you need anything more from me, otherwise I'll wait for the next patch that addresses ^^

ramaraochavali commented:
@htuch since we are good with the proto, i will add few more tests and address the other review comments and let you know.

ramaraochavali commented on api/envoy/admin/v2alpha/config_dump.proto:
```diff
@@ -178,3 +182,38 @@ message RoutesConfigDump {
+    // The xDS service this control plane information belongs to. For example
+    // envoy.api.v2.EndpointDiscoveryService represents control plane info that Envoy has been
+    // configured for EndpointDiscoveryService.
+    string service = 1;
```
Yes. I can change it to `server_type_url`

ramaraochavali commented:
@htuch addressed the review comments and added an integration test in cds. Please take a look. Let me know if we need any additional tests - I was thinking of adding a test that captures two clusters pointing to two different gRPC services for EDS - but seems little difficult to add with current setup. Do you have any suggestions to add it with reasonable effort?

htuch commented on source/common/config/grpc_mux_impl.h:
```diff
@@ -89,6 +94,17 @@ class GrpcMuxImpl
+  // Per Service Control Plane configuration that Envoy is connected to. For services like
+  // RouteDiscoveryService and EndpointDiscovery the configuration allows to specify multiple config
+  // sources, for example a different RouteDiscoveryService can be specified for each listener.
+  static std::unordered_map<
```
Static non-PODs are disallowed due to the static initialization fiasco. I'm wondering if static is really the right thing here, since it also might be harder from a testing perspective to reason about. Is there an alternative that isn't too hard to implement instead?

ramaraochavali replied:
I agree it is harder to reason about and test. But could not find any better alternative. Let me know if you have any ideas - I can implement it.

htuch replied:
Hmm. Can we make `ConfigTracker` smarter somehow? This is essentially responsible for keeping track of all the configs and aggregation, so it seems a natural place to consider doing this. And/or change the shape of the proto data model to make this easier.

ramaraochavali replied:
Thinking more about this, I can think of following options
1. Make `ConfigTracker` smart by introducing some thing similar to `managed_config` which is actually managed by `ConfigTracker` instance - so if this config is declared as `managed_config`, whenever this needs to be changed we can get, update and set it back. When '/config_dump' is called, the `ConfigTracker` can return the config by itself for managed configs instead of calling config dump function call back.
2. Can we just store this specific config in `Singleton Manager` avoiding static here?

Option 2 may be simpler to implement, Option 1 might be a useful addition to `ConfigTracker` framework and can be used in future.

WDYT?

htuch replied:
What do you mean by `managed_config` in (1)? I think in general, putting the smarts in `ConfigTracker` is the best way to go, so +1.

ramaraochavali replied:
By `managed_config` - I mean the config dump proto structure is owned by `ConfigTracker` and `grpc_mux_impl` would get and update it whenever it needs and when `/config_dump` is called for this type `managed_config` the `ConfigTracker` would just dump what it has instead of calling the callback.

htuch replied:
Yep, that sounds like a good solution.

htuch commented on api/envoy/admin/v2alpha/config_dump.proto:
```diff
@@ -178,3 +183,38 @@ message RoutesConfigDump {
+    // The xDS service this control plane information belongs to. For example
+    // envoy.api.v2.EndpointDiscoveryService represents control plane info that Envoy has been
+    // configured for EndpointDiscoveryService.
+    string server_type_url = 1;
```
Actually, this isn't type URL, it's the fully qualified service name, so maybe call it `xds_service`. Sorry about the confusion.

htuch commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -206,5 +220,47 @@ void GrpcMuxImpl::handleEstablishmentFailure() {
+        break;
+      }
+    }
+    config_source_control_plane_info.mutable_control_plane()->set_identifier(
```
Can you add a few more comments to this method? I think it's correct but it took me a while to figure out the different cases.

htuch commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -206,5 +220,47 @@ void GrpcMuxImpl::handleEstablishmentFailure() {
+        service_control_plane_info;
+    service_control_plane_info.set_server_type_url(service_info.first);
+    for (auto& config_source_control_plane_info : service_info.second) {
+      std::string iden = config_source_control_plane_info.control_plane().identifier();
```
Where is this used?

htuch commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -206,5 +220,47 @@ void GrpcMuxImpl::handleEstablishmentFailure() {
+      GrpcMuxImpl::per_service_control_plane_info_[server_type_url_].push_back(
+          config_source_control_plane_info);
+    }
+  }
```
Is it possible that for the same service, e.g. EDS, we might have two identical `GrpcService` configurations, but they connect to distinct servers which return distinct identifiers in their `DiscoveryResponse`?

ramaraochavali replied:
Yes. It is possible and in that case, it goes in to update mode and updates the control_plane_identifier for existing gRPC service.

htuch replied:
if cluster X points to GrpcService M for EDS and cluster Y also points to M, In the absence of https://github.com/envoyproxy/envoy/issues/2943, you might be speaking to two different servers at the same time. You probably don't want update semantics in this case.

ramaraochavali replied:
Hmm.. How do we distinguish that case now? Do we have enough information or should we also include the cluster in the mix (that may be too much)?

htuch replied:
I think fundamentally you do need the cluster/listener there, or we need #2943 to land for EDS (as well as every other API that might have this property, e.g. RDS, SDS).

An alternative that might be simpler is making it a set (unordered list) of identifiers, but you need some way internally of knowing when to delete and modify the identifiers, so need some similar kind of association.

ramaraochavali commented:
I am busy with other things and could not get it to implementing some of the changes. Will do.

ramaraochavali commented:
/no stale
